### PR TITLE
Add vtgate and vttablet flag to truncate errors sent to client

### PR DESCRIPTION
**Pull request by dasl-**
## Description

Hi there, I'm reviving this old PR which got closed: https://github.com/vitessio/vitess/pull/6916 . That old PR added a flag to truncate long errors on vttablet (`-queryserver-config-truncate-error-len`). 

This PR adds the same vttablet flag to truncate long errors, and additionally, it adds a similar flag for vtgate. This PR introduces a pair of flags, one for vttablet and one for vtgate:

- vttablet: `-queryserver-config-truncate-error-len`
- vtgate: `-truncate-error-len`

_Note for reviewers:_ you can browse the two commits in this PR separately to see the changes related to adding each flag in isolation.

Unlike vanilla MySQL, vttablet sometimes includes client SQL in error responses leading to potentially large response packets. This confuses some clients such as PHP's PDO driver. (A bug has been filed with PHP[0] for this behavior.)

Similarly, vtgate can also send large error messages to the client for errors that originate in vtgate.

We avoid this edge case by truncating large error messages via these flags.

[0] https://bugs.php.net/bug.php?id=78797

## Related Issue(s)

Fixes https://github.com/vitessio/vitess/issues/13043

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on the CI
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/flags/endtoend/vttablet.txt [hunk 1]
```diff
@@ -236,6 +236,7 @@ Usage of vttablet:
       --queryserver-config-terse-errors                                  prevent bind vars from escaping in client error messages
       --queryserver-config-transaction-cap int                           query server transaction cap is the maximum number of transactions allowed to happen at any given point of a time for a single vttablet. E.g. by setting transaction cap to 100, there are at most 100 transactions will be processed by a vttablet and the 101th transaction will be blocked (and fail if it cannot get connection within specified timeout) (default 20)
       --queryserver-config-transaction-timeout float                     query server transaction timeout (in seconds), a transaction will be killed if it takes longer than this value (default 30)
+      --queryserver-config-truncate-error-len int                        truncate errors sent to client if they are above this value (0 means do not truncate)
```
Comment (shlomi-noach) [inline go/flags/endtoend/vttablet.txt:239]: Does this rewrite make sense?
```suggestion
      --queryserver-config-truncate-error-len int                        truncate errors sent to client if they are longer than this value (0 means do not truncate)
```

File: go/flags/endtoend/vtgate.txt [hunk 2]
```diff
@@ -199,6 +199,7 @@ Usage of vtgate:
       --vmodule moduleSpec                                               comma-separated list of pattern=N settings for file-filtered logging
       --vschema_ddl_authorized_users string                              List of users authorized to execute vschema ddl operations, or '%' to allow all users.
       --vtgate-config-terse-errors                                       prevent bind vars from escaping in returned errors
+      --vtgate-config-truncate-error-len int                             truncate errors sent to client if they are above this value (0 means do not truncate)
```
Comment (shlomi-noach) [inline go/flags/endtoend/vtgate.txt:202]: Does this rewrite make sense?
```suggestion
      --vtgate-config-truncate-error-len int                             truncate errors sent to client if they are longer than this value (0 means do not truncate)
```

File: go/vt/vtgate/vtgate.go [hunk 3]
```diff
@@ -117,6 +118,7 @@ func registerFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&transactionMode, "transaction_mode", transactionMode, "SINGLE: disallow multi-db transactions, MULTI: allow multi-db transactions with best effort commit, TWOPC: allow multi-db transactions with 2pc commit")
 	fs.BoolVar(&normalizeQueries, "normalize_queries", normalizeQueries, "Rewrite queries with bind vars. Turn this off if the app itself sends normalized queries with bind vars.")
 	fs.BoolVar(&terseErrors, "vtgate-config-terse-errors", terseErrors, "prevent bind vars from escaping in returned errors")
+	fs.IntVar(&truncateErrorLen, "vtgate-config-truncate-error-len", truncateErrorLen, "truncate errors sent to client if they are above this value (0 means do not truncate)")
```
Comment (shlomi-noach) [inline go/vt/vtgate/vtgate.go:121]:
```suggestion
	fs.IntVar(&truncateErrorLen, "vtgate-config-truncate-error-len", truncateErrorLen, "truncate errors sent to client if they are longer than this value (0 means do not truncate)")
```

File: go/vt/vttablet/tabletserver/tabletenv/config.go [hunk 4]
```diff
@@ -131,6 +131,7 @@ func registerTabletEnvFlags(fs *pflag.FlagSet) {
 	fs.BoolVar(&currentConfig.EnableTableACLDryRun, "queryserver-config-enable-table-acl-dry-run", defaultConfig.EnableTableACLDryRun, "If this flag is enabled, tabletserver will emit monitoring metrics and let the request pass regardless of table acl check results")
 	fs.StringVar(&currentConfig.TableACLExemptACL, "queryserver-config-acl-exempt-acl", defaultConfig.TableACLExemptACL, "an acl that exempt from table acl checking (this acl is free to access any vitess tables).")
 	fs.BoolVar(&currentConfig.TerseErrors, "queryserver-config-terse-errors", defaultConfig.TerseErrors, "prevent bind vars from escaping in client error messages")
+	fs.IntVar(&currentConfig.TruncateErrorLen, "queryserver-config-truncate-error-len", defaultConfig.TruncateErrorLen, "truncate errors sent to client if they are above this value (0 means do not truncate)")
```
Comment (shlomi-noach) [inline go/vt/vttablet/tabletserver/tabletenv/config.go:134]:
```suggestion
	fs.IntVar(&currentConfig.TruncateErrorLen, "queryserver-config-truncate-error-len", defaultConfig.TruncateErrorLen, "truncate errors sent to client if they are longer than this value (0 means do not truncate)")
```

**Review by shlomi-noach — COMMENTED** [review]
Just minor nits about phrasing. Overall this LGTM, but I'll defer to other reviewers

File: go/flags/endtoend/vtgate.txt [hunk 5]
```diff
@@ -199,6 +199,7 @@ Usage of vtgate:
       --vmodule moduleSpec                                               comma-separated list of pattern=N settings for file-filtered logging
       --vschema_ddl_authorized_users string                              List of users authorized to execute vschema ddl operations, or '%' to allow all users.
       --vtgate-config-terse-errors                                       prevent bind vars from escaping in returned errors
+      --vtgate-config-truncate-error-len int                             truncate errors sent to client if they are longer than this value (0 means do not truncate)
```
[the remark this answers is not in the document]
Comment (dasl-) [reply go/flags/endtoend/vtgate.txt]: Thanks for the suggestion - I've updated the PR now.

**Review by harshit-gangal — COMMENTED** [review]
How about we add `...` or `<truncated>` at the end to indicate the truncation of the message.

**Comment by systay** [comment]
I've not followed this discussion in the past, but I'm just thinking - do we really want two different flags for this? Shouldn't we use the same flag for both binaries? Am I missing something?

**Comment by harshit-gangal** [comment]
> I've not followed this discussion in the past, but I'm just thinking - do we really want two different flags for this? Shouldn't we use the same flag for both binaries? Am I missing something?

vtcombo (vtgate + vttablet single binary) will panic if you use same name for two flags

**Comment by dasl-** [comment]
> How about we add ... or <truncated> at the end to indicate the truncation of the message.

@harshit-gangal Good suggestion, I've updated with the changes now.

> I've not followed this discussion in the past, but I'm just thinking - do we really want two different flags for this? Shouldn't we use the same flag for both binaries? Am I missing something?

@systay I'm happy to change the naming of the flags. But it sounds like there's some constraints at the moment with respect to unique flag names in vtgate vs vttablet. Do you think we should block merging this PR until the work can be done to have unique flag names?

**Review by shlomi-noach — APPROVED** [review]
As a PR requires two approvals, I'll go ahead and give one; we'll wait on @systay decision on how to resolve the command line flags issue.

**Comment by harshit-gangal** [comment]
@dasl- Vitess has `truncateSQLAndBindVars` implementation already which will reduce the SQL length added in the error message sent to client that is controlled using `sql-max-length-errors` flag. 

`truncateSQLAndBindVars` is used in returning error from vttablet to vtgate.

I think this should be enough for your use case. Let me know what you think about it.

If needed we can use the same method in vtgate as well.

**Comment by dasl-** [comment]
@harshit-gangal Thanks for your suggestion. Unfortunately that won't work in all cases: sometimes a long error message can be generated that is unrelated to a long SQL query. For instance, in a keyspace with many shards, the `multi-db transaction` error can be quite long:

```
multi-db transaction attempted: [target:{keyspace:\"etsy_shard\" shard:\"3667-36ab\" tablet_type:PRIMARY} transaction_id:1679515688995875481 tablet_alias:{cell:\"us_central1_a\" uid:1664531293} target:{keyspace:\"etsy_shard\" shard:\"d445-d489\" tablet_type:PRIMARY} transaction_id:1679515341617366636 tablet_alias:{cell:\"us_central1_b\" uid:3200715491} target:{keyspace:\"etsy_shard\" shard:\"5aef-5b34\" tablet_type:PRIMARY} transaction_id:1679515125918994230 tablet_alias:{cell:\"us_central1_f\" uid:649326074} target:{keyspace:\"etsy_shard\" shard:\"4a23-4a67\" tablet_type:PRIMARY} transaction_id:1679515030389209428 ... etc ...
```

https://github.com/vitessio/vitess/blob/c5dced7f4527a064c066a123c5b897be7f1ce3ec/go/vt/vtgate/safe_session.go#L445

This error contains no SQL text.

**Comment by harshit-gangal** [comment]
> @harshit-gangal Thanks for your suggestion. Unfortunately that won't work in all cases: sometimes a long error message can be generated that is unrelated to a long SQL query. For instance, in a keyspace with many shards, the `multi-db transaction` error can be quite long:
> 
> ```
> multi-db transaction attempted: [target:{keyspace:\"etsy_shard\" shard:\"3667-36ab\" tablet_type:PRIMARY} transaction_id:1679515688995875481 tablet_alias:{cell:\"us_central1_a\" uid:1664531293} target:{keyspace:\"etsy_shard\" shard:\"d445-d489\" tablet_type:PRIMARY} transaction_id:1679515341617366636 tablet_alias:{cell:\"us_central1_b\" uid:3200715491} target:{keyspace:\"etsy_shard\" shard:\"5aef-5b34\" tablet_type:PRIMARY} transaction_id:1679515125918994230 tablet_alias:{cell:\"us_central1_f\" uid:649326074} target:{keyspace:\"etsy_shard\" shard:\"4a23-4a67\" tablet_type:PRIMARY} transaction_id:1679515030389209428 ... etc ...
> ```
> 
> https://github.com/vitessio/vitess/blob/c5dced7f4527a064c066a123c5b897be7f1ce3ec/go/vt/vtgate/safe_session.go#L445
> 
> This error contains no SQL text.

This can happen at VTGate only. So, I will suggest removing changes from vttablet and using the already present truncation logic there with the correct flag.
For vtgate, we should not do the changes in the MySQL package. It does not feel correct to do any logic changes around truncation there. It should be moved to `vtgate/executor`.

**Comment by dasl-** [comment]
@harshit-gangal I was taking another look at the `sql-max-length-errors` you mentioned:
```
--sql-max-length-errors int                                        truncate queries in error logs to the given length (default unlimited)
```

I believe this flag only applies to long queries that are logged to disk. I'm not sure that it also applies to errors returned to the client. Thus, I think we might still need the flags added in this PR, which specifically truncate errors sent to the client. I suppose another option is extending the `sql-max-length-errors` flag to apply to both use cases though.

Furthermore, the truncation given by the `sql-max-length-errors` flag is not always applied where it should be. See [this PR](https://github.com/vitessio/vitess/pull/12944) where I am trying to apply the truncation in various spots where it appears to have been missed. To some extent, it seems like a game of "whack-a-mole" trying to cover all callsites with the `sql-max-length-errors` flag.

Since all errors appear to be emitted by the same codepath ([here in vttablet](https://github.com/vitessio/vitess/blob/966394d4af29d433a1b66bddf91d2a52231a9c9f/go/vt/vttablet/tabletserver/tabletserver.go#L1500) and [here in vtgate](https://github.com/vitessio/vitess/blob/966394d4af29d433a1b66bddf91d2a52231a9c9f/go/mysql/conn.go#L843)), it seems like a more foolproof method of ensuring errors returned to clients are not too long. 

>For vtgate, we should not do the changes in the MySQL package. It does not feel correct to do any logic changes around truncation there. It should be moved to `vtgate/executor`.

I'm curious where you would recommend placing this logic in `vtgate/executor`. I assume you were suggesting I move this logic: https://github.com/vitessio/vitess/blob/966394d4af29d433a1b66bddf91d2a52231a9c9f/go/mysql/conn.go#L845-L847

It's not obvious to me where that would go in `vtgate/executor`. Thanks for your feedback!

File: go/vt/vterrors/vterrors.go [hunk 6]
```diff
@@ -361,5 +361,19 @@ func Print(err error) string {
...
+		return msg
+	}
+
+	if max <= 12 {
+		return "[TRUNCATED]"
+	}
+
+	return msg[:max-12] + " [TRUNCATED]"
+}
+
```
Comment (harshit-gangal) [inline go/vt/vterrors/vterrors.go:377]: I think this method itself should return the new error with truncated error message.

**Comment by harshit-gangal** [comment]
I see that in tabletserver. The tabletserver side changes looks good.

In VTGate, executor side needs to be change in `Prepare`, `Execute`, `StreamExecute` methods

**Comment by dasl-** [comment]
@harshit-gangal I think this is ready for another look. I believe the failing `onlineddl_vrepl` test is due to a flakey timeout.

**Review by systay — APPROVED** [review]
> @systay I'm happy to change the naming of the flags. But it sounds like there's some constraints at the moment with respect to unique flag names in vtgate vs vttablet. Do you think we should block merging this PR until the work can be done to have unique flag names?

No, you are right, we should fix that separately.

File: go/vt/vterrors/vterrors.go [hunk 7]
```diff
@@ -361,5 +361,19 @@ func Print(err error) string {
...
+func TruncateError(oldErr error, max int) error {
+	if oldErr == nil || max <= 0 || len(oldErr.Error()) <= max {
+		return oldErr
+	}
+
+	if max <= 12 {
+		return New(Code(oldErr), "[TRUNCATED]")
+	}
+
+	return New(Code(oldErr), oldErr.Error()[:max-12]+" [TRUNCATED]")
```
[the remark this answers is not in the document]
Comment (harshit-gangal) [reply go/vt/vterrors/vterrors.go:375]: looks like it, should move to previous version then.

File: go/vt/vtgate/vtgate.go [hunk 8]
```diff
@@ -304,6 +306,7 @@ func Init(
 		si,
 		noScatter,
 		pv,
+		truncateErrorLen,
```
Comment (harshit-gangal) [inline go/vt/vtgate/vtgate.go:309]: `truncateErrorLen` does not need to be part of the parameter. `vtgate` and `executor` both have access to this field.

Comment (dasl-) [reply go/vt/vtgate/vtgate.go]: good catch; updated!

File: go/vt/vtgate/executor.go [hunk 9]
```diff
@@ -134,20 +138,22 @@ func NewExecutor(
...
+		scatterConn:      resolver.scatterConn,
+		txConn:           resolver.scatterConn.txConn,
+		plans:            cache.NewDefaultCacheImpl(cacheCfg),
+		normalize:        normalize,
+		warnShardedOnly:  warnOnShardedOnly,
+		streamSize:       streamSize,
+		schemaTracker:    schemaTracker,
+		allowScatter:     !noScatter,
+		pv:               pv,
+		truncateErrorLen: truncateErrorLen,
```
[the remark this answers is not in the document]
Comment (dasl-) [reply go/vt/vtgate/executor.go]: good catch; updated!

**Review by deepthi — DISMISSED** [review]
@dasl- we've started requiring GH issues for all bugs and new features.
Can you please create an issue that just describes the desired feature (no implementation details) and link it here?

**Comment by dasl-** [comment]
@deepthi sure, I've filed this issue: https://github.com/vitessio/vitess/issues/13043

**Comment by dasl-** [comment]
FYI I've added website documentation for these new flags in this PR: https://github.com/vitessio/website/pull/1459

**Review by harshit-gangal — APPROVED** [review]

**Comment by harshit-gangal** [comment]
@dasl- can you resolve the conflict here

**Comment by dasl-** [comment]
@harshit-gangal I believe I have resolved the merge conflict. Looks like some tests are being flakey and failing right now, but I _think_ this should be good to merge if the tests can be re-run.

**Pull request merged**